File: kickstart/tmp/redo-DA.py
def isOdd(n):
    return n % 2


def DA_solve(num_list, O, D):
    L = 0
    R = 0
    length = len(num_list)
    maxsum = -99999
    sums = [0 for i in range(length)]
    odds = [0 for i in range(length)]

    odds[0] = isOdd(num_list[0])
    sums[0] = num_list[0]
    for i in range(1, length):
        odds[i] = odds[i - 1] + isOdd(num_list[i])
        sums[i] = sums[i - 1] + num_list[i]

    def os(l, r):
        if 0 < l:
            totalodd = odds[r] - odds[l - 1]
            return totalodd
        else:
            return odds[r]

    def ss(l, r):
        if 0 < l:
            totals = sums[r] - sums[l - 1]
            return totals
        else:
            return sums[r]

    for R in range(length):
        if L > R:
            continue

        while os(L, R) > O or ss(L, R) > D:
            L += 1
            if L > R:
                break

        if L > R:
            continue

        tmpsum = ss(L, R)
        if tmpsum > maxsum:
            maxsum = tmpsum

    # while R < length:
    #     while os(L, R) > O:
    #         L += 1
    #         if L>R:
    #             R+=1
    #             break
    #     #then below will be os(L, R) <= O

    #     while ss(L, R) > D:
    #         L += 1
    #         if L>R:
    #             R+=1
    #             break
    #     #then below will be ss(L, R) <= D

    #     tmpsum = ss(L, R)
    #     if tmpsum > maxsum:
    #         maxsum = tmpsum

    #     R += 1

    if maxsum == -99999 or maxsum > D:
        print('IMPOSSIBLE')
    else:
        print(maxsum)

    return 0


def main():
    T = int(input())

    for n in range(T):
        tmp = input().split()
        tmp = list(map(int, tmp))
        N, O, D = tmp

        tmp = input().split()
        tmp = list(map(int, tmp))
        x1, x2, A, B, C, M, L = tmp

        X = [x1, x2]
        S = [x1 + L, x2 + L]

        for i in range(2, N):
            X.append((A * X[i - 1] + B * X[i - 2] + C) % M)
            S.append(X[i] + L)

        print('Case #' + str(n + 1) + ': ', end='')
        DA_solve(tuple(S), O, D)


if __name__ == '__main__':
    main()

File: kickstart/tmp/redo-DA.cpp
#include<stdio.h>
#include<vector>

const int MAXN = 500005;

int X[MAXN];
int S[MAXN];
long long sum[MAXN];

int isOdd(int n){
    return n % 2;
}

int os(vector & num_list,int l,int r){
        if (l > r){
            return 0;
        }
        int totalodd = 0;
        for(int i=l;i<r+1;i++){
            if (isOdd(num_list[i])){
                totalodd += 1;
            }
        }
        return totalodd    
}

int ss(vector &sums,int l,int r){
        if (l > r){
            return 0;
        }
        totals = sums[l]-sums[r-1]
        return totals

}
    
int DA_solve(vector &num_list,int O,int D){
    int L = 0;
    int R = 0;
    int length = num_list.length;
    long long maxsum = -0x3f3f3f3f3f3f3f3fLL;

    
    while (R < length){
        while (os(L, R) > O && L < R){
            L += 1;
        }

        while (ss(L, R) > D && L < R){
            L += 1;
        }

        tmpsum = ss(L, R);
        if (tmpsum > maxsum){
            maxsum = tmpsum;
        }

        R += 1;

    }
        
    if (maxsum == -99999 || maxsum > D){

        printf("IMPOSSIBLE\n");
    }
    else:
        printf("%d\n",maxsum);

    return 0;


}
    
int main(){
    int T = 0;
    int N,O,D;
    int x1, x2, A, B, C, M, L;
    scanf('%d',&T);

    for (int i = 0; i < T; ++i)
    {   
        scanf("%d %d %d",&N,&O,&P);
        scnaf("%d %d %d %d %d %d %d",&x1, &x2, &A, &B, &C, &M, &L);
        X[0]=x1;X[1]=x2;
        for (int i = 3; i <= N; i++) {
        X[i] = (1LL * A * X[i - 1] + 1LL * B * X[i - 2] + C) % M;
        }
        for (int i = 1; i <= N; i++) {
            S[i] = X[i] + L;
            sum[i] = sum[i - 1] + S[i];
        }
    }

    for n in range(T):
        tmp = input().split()
        tmp = list(map(int, tmp))
        N, O, D = tmp

        tmp = input().split()
        tmp = list(map(int, tmp))
        x1, x2, A, B, C, M, L = tmp

        X = [x1, x2]
        S = [x1 + L, x2 + L]

        for i in range(2, N):
            X.append((A * X[i - 1] + B * X[i - 2] + C) % M)
            S.append(X[i] + L)

        print('Case #' + str(n + 1) + ': ', end='')
        DA_solve(tuple(S), O, D)

}
    